internal/app/db.go

```diff
@@ -0,0 +1,316 @@
+		uploader:  uploader,
+	}
+}
+
+// Close closes the current db.
+func (dbm *DBManager) Close() error {
+	return dbm.db.Close()
+}
+
+// queryFromWAL creates a query for a WAL TX records.
+func (dbm *DBManager) queryFromWAL(tx *pgrepl.Tx) string {
+	var queries []string
+	for _, r := range tx.Records {
+		columnNames := []string{}
+		vals := []string{}
+		for _, c := range r.Columns {
+			columnNames = append(columnNames, c.Name)
+			vals = append(vals, strings.ReplaceAll(string(c.Value), "\"", ""))
+		}
+
+		cols := strings.Join(columnNames, ", ")
+		valsStr := strings.Join(vals, ", ")
+		query := fmt.Sprintf(
+			"insert into %s (%s) values (%s) \n",
+			r.Table, cols, valsStr,
+		)
+		queries = append(queries, query)
+	}
+
+	// batch all insert queries into one
+	return strings.Join(queries, "\n")
+}
+
+func (dbm *DBManager) replace(ctx context.Context) error {
+	// Export current db to a parquet file
+	now := time.Now()
+	exportPath := path.Join(dbm.dbDir, fmt.Sprintf("%d.db", now.UnixNano()))
+	if err := dbm.Export(ctx, exportPath); err != nil {
+		return err
+	}
+
+	// Close current db
+	slog.Info("closing current db")
+	if err := dbm.Close(); err != nil {
+		return err
+	}
+
+	// Upload the exported parquet file
+	if err := dbm.UploadAt(ctx, exportPath); err != nil {
+		fmt.Println("upload error, skipping", "err", err)
+	}
+
+	// Create a new db
+	db, err := dbm.NewDB()
+	if err != nil {
+		return fmt.Errorf("new db: %v", err)
+	}
+
+	// Setup the new db
+	dbm.db = db
+	dbm.createdAT = now
+	return dbm.Setup(ctx)
+}
+
+// NewDB creates a new duckdb database at the current.db path.
+func (dbm *DBManager) NewDB() (*sql.DB, error) {
+	now := time.Now()
+	dbPath := path.Join(dbm.dbDir, fmt.Sprintf("%d.db", now.UnixNano()))
+	db, err := sql.Open("duckdb", dbPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open db: %s", err)
+	}
+
+	return db, nil
+}
+
+// Setup creates a local table in the current.db and installs
+// the parquet extension.
+func (dbm *DBManager) Setup(ctx context.Context) error {
+	query, err := dbm.genCreateQuery()
+	if err != nil {
+		return fmt.Errorf("cannot create table in duckdb: %s", err)
+	}
+
+	// create table if it does not exist
+	slog.Info("applying create", "query", query)
+	_, err = dbm.db.ExecContext(ctx, query)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (dbm *DBManager) windowPassed() bool {
+	if dbm.createdAT.IsZero() {
+		return false
+	}
+	delta := time.Since(dbm.createdAT).Seconds()
+	threshold := dbm.winSize.Seconds()
+	return delta > threshold
+}
+
+// Replay replays a WAL record onto the current db.
+// If the window has passed, it replaces the current db with
+// a new one. The current db is exported and uploaded before
+// new db is ready to be used.
+func (dbm *DBManager) Replay(ctx context.Context, tx *pgrepl.Tx) error {
+	if dbm.windowPassed() {
+		slog.Info("replacing current db before replaying further txs")
+		if err := dbm.replace(ctx); err != nil {
```

Comment:
If export or upload fails, the replicator will exit. the next time it starts it will first upload all existing db files and then start replicating on a fresh db